README.md

```diff
@@ -1,2 +1,1088 @@
+`add_help`, and `allow_abbrev`.
+
+The epilog is always appended with information about the origins of the tool.
+For example, the [*Maze Editor*](#maze-editor)'s epilog states it was loaded
+from the `MazeEditor` class in the `sim.tools` module.
+
+The tool's class must override the `build_parser(...)` class method that
+accepts the tool's (`argparse`) parser and adds arguments to it, and the
+`main(...)` class method that accepts the `argparse.Namespace` object that
+returned from the `parse_args(...)` call and performs the tool's logic.
+
+For an example of a simple tool see the [*Maze Editor*](#maze-editor) tool's
+[implementation](./sim/tools.py).
+
+## Maze Formats
+
+This package supports 4 different maze formats.
+These formats are commonly used in other Micromouse-related projects.
+
+### The ".maze" Format
+
+This format is essentially an ASCII-art drawing of the maze but any UTF-8
+character is allowed.
+
+Cells are `5x3` by default and share walls in the drawing (for example, a `2x3`
+maze will take `5x13` characters).
+However, only the center of every wall is checked (by default against a space
+character: `' '`).
+
+For example:
+
+```sh
+$ cat mazes/example.maze
++---+---+---+
+|       |   |
++   +   +   +
+|   |       |
++---+---+---+
+```
+
+This is the default format.
+
+### The ".maz" Format
+
+This format uses bit flags for the walls of each cell.
+
+```plain
+0x1 - North
+0x2 - East
+0x4 - South
+0x8 - West
+```
+
+An entire byte is saved per  although only a nibble per cell would suffice.
+
+The maze is represented as a row stack, meaning that for an `nxm` maze, the
+first byte is the `(0, 0)` cell, the second `(0, 1)`, the cell at index `m` is
+`(1, 0)` and so on.
+
+Since this is a binary format and cannot communicate maze boundaries, the
+dimensions of the maze should be communicated in the filename suffix:
+`.{height}x{width}.maz`. If dimensions are not specified but the content's
+length (the number of cells) is an integer square, the maze is assumed to be a
+square maze of that size.
+
+For example:
+
+```sh
+$ hexdump -C ./mazes/example.2x3.maz
+00000000  09 03 0b 0e 0c 06                                 |......|
+00000006
+$ micromouse tool maze render ./mazes/example.2x3.maz
++---+---+---+
+|       |   |
++   +   +   +
+|   |       |
++---+---+---+
+```
+
+### The ".num" Format
+
+This format is a simple text-based format where each line is in the format
+`X Y N E S W` - 6 integers that represent:
+
++ `X` & `Y` - The cell's XY coordinates. Where `(0, 0)` is the ***bottom-left***
+  corner (not the *top-left* corner as used everywhere else in this project).
++ `N`, `E`, `S`, `W` - North, East, South, West - 4 booleans (`0`/`1`)
+  representing where the cell has walls.
+
+The `Y` coordinate is inverted because the format is based on
+[this simulator](https://github.com/mackorone/mms) that uses `(0, 0)` as the
+bottom-left corner rather than `(0, 0)` as in this project.
```

Comment:
Not sure if I like that repeating of `(0, 0)` without saying how we defined it in this project.
But maybe it's just me...

Reply:
Yeah that said "top-left" and I'm not sure what undo/copy-paste caused this